File: typescript/LWTSD/DataController.ts
var xmpp = require('../node-xmpp/packages/client').xmpp;
var xml = require('../node-xmpp/packages/xml');
var jid = require('../node-xmpp/packages/jid');
var Guid = require('../Misc/Guid');

import {LWTSD} from './LWTSD';
import {SchemaFormat, ParseSchemaFormat, SchemaFormatToString} from './SchemaFormat';
import {SimplifiedSchema} from './SimplifiedSchema'
import {SimplifiedType} from './SimplifiedType'
import {Resource} from './ResourceTypes/Resource'

import {ResourceBase64Binary} from './ResourceTypes/ResourceBase64Binary';
import {ResourceBoolean} from './ResourceTypes/ResourceBoolean';
import {ResourceDateTime} from './ResourceTypes/ResourceDateTime';
import {ResourceDecimal} from './ResourceTypes/ResourceDecimal';
import {ResourceDouble} from './ResourceTypes/ResourceDouble';
//import {ResourceDuration} from './ResourceTypes/ResourceDuration';
import {ResourceFloat} from './ResourceTypes/ResourceFloat';
import {ResourceInteger} from './ResourceTypes/ResourceInteger';
import {ResourceString} from './ResourceTypes/ResourceString';
import { NewGuid } from '../Misc/Guid';
//import {ResourceTime} from './ResourceTypes/ResourceTime';


class SchemaMissMatchException
{
}

export class DataPage
{
	public static readonly PointsPerPage : number = 500;
	
	public Data : Array<Resource> = new Array<Resource>();
	public Schema : SimplifiedSchema |null = null;
	public NrTotalPages : number = 1;
	public Page : number = 0;

	public constructor(Element? : any, KnownSchema? : SimplifiedSchema, Page? : number)
	{
		if(!Element)
			return;

		if (KnownSchema == null)
			throw new SchemaMissMatchException();

		let NewSchemaVersion = Element.attrs.schemaversion;
		if (NewSchemaVersion != KnownSchema.GetVersion())
			throw new SchemaMissMatchException();

		this.Schema = KnownSchema;
		this.NrTotalPages = Math.round((parseInt(Element.attrs.totalpoints)+ 0.5)/DataPage.PointsPerPage);

		let resources = Element.getChildren("resource");

		for(let i = 0; i < resources.length; i++)
		{
			let it = resources[i];
			let Path = it.attrs.path;

			let rdesc = KnownSchema.Resources.get(Path);
			if(!rdesc)
				continue;

			switch (rdesc.SimpleType)
			{
				case SimplifiedType.String:
					this.Data.push(new ResourceString(it));
					break;
				case SimplifiedType.Integer:
					this.Data.push(new ResourceInteger(it));
					break;
				case SimplifiedType.Base64Binary:
					this.Data.push(new ResourceBase64Binary(it));
					break;
				case SimplifiedType.DateTime:
					this.Data.push(new ResourceDateTime(it));
					break;
				case SimplifiedType.Boolean:
					this.Data.push(new ResourceBoolean(it));
					break;
				case SimplifiedType.Decimal:
					this.Data.push(new ResourceDecimal(it));
					break;
				case SimplifiedType.Duration:
					// TODO this.Data.push(new ResourceDuration(it));
					break;
				case SimplifiedType.Float:
					this.Data.push(new ResourceFloat(it));
					break;
				case SimplifiedType.Time:
					// TODO this.Data.push(new ResourceTime(it));
					break;
			}
		}
	}
}


class Subscription
{
	public ID : string;
	public NewData : (ID : string, Data :  DataPage) => void;
	public SubscriptionCancelled : (ID : string) => void;
	public Resources : Array<string>;
	public KnownSchema : SimplifiedSchema | null;
	public AccessToken : string | null;
	public LastReSubscription : Date = new Date();
}


export class DataController
{
	private DataSourceAddress : string; // JID
	private Uplink : any; // xmp

	private IQResolvers: Map<string, (IQElement : any) => void> = new  Map<string, (IQElement : any) => void> ();
	private IQRejecters: Map<string, (IQElement : any) => void> = new  Map<string, (IQElement : any) => void> ();

	private Subscriptions : Map<string, Subscription> = new Map<string, Subscription>();

	private OnStanzaIQ(iq :any) : void
	{
		try
        {
			if(iq.attrs.from != this.DataSourceAddress)			
	        {
	            console.log("Warning; got iq from not DataSource? From = " + iq.attrs.from);
	            return;
	        }


	        if(this.IQResolvers.has(iq.attrs.id))
	        {
	            console.log("Running resolver on " + iq.attrs.id);
	            let resolver = this.IQResolvers.get(iq.attrs.id);
	            if(resolver)
	                resolver( iq );                
	            this.IQResolvers.delete(iq.attrs.id);
	            this.IQRejecters.delete(iq.attrs.id);
	            return;
	        }

	        console.log("Warning; got iq from unregistered request from " + iq.attrs.from + " with id  = " + iq.attrs.id);
	    }
	    catch(err)
	    {
	        console.log("Unhandled exception i DataController: " + err);
	    }
    
	}

	private OnStanzaMessage(Message : any) : void
	{
		try
        {
			if(Message.attrs.from != this.DataSourceAddress)			
	        {
	            console.log("Warning; got message from not DataSource? From = " + Message.attrs.from);
	            return;
	        }

			let LWTSDElement = Message.getChildren()[0];

			if (LWTSDElement.getNS() != LWTSD.Namespace)
				return;

			switch (LWTSDElement.name)
			{
				case "subscription-cancelled":
					this.HandleSubscriptionCancelled(LWTSDElement);
					break;
				case "subscription-triggered":
					this.HandleSubscriptionTriggered(LWTSDElement);
					break;
			}

	    }
	    catch(err)
	    {
	        console.log("Unhandled exception i DataController: " + err);
	    }	
	}

 	private OnStanza ( Message : any) : void
    {
    	if(Message.is('iq'))
    		this.OnStanzaIQ(Message);

    	if(Message.is('message'))
    		this.OnStanzaMessage(Message);
    }

    private async GetIQResponse(Request :any) 
    : Promise<any>
    {
    	let Resolves = this.IQResolvers;
		let Rejects = this.IQRejecters;
		let Uplink = this.Uplink;

		let TProm = new Promise<any>(
			function(resolve, reject)
			{
				Resolves.set(Request.attrs.id, resolve);
				Rejects.set(Request.attrs.id, reject);
				Uplink.send(Request);
			});

		return TProm;

    }

	public async  GetSchema(Page? :number, 
		AccessToken? :string |null, 
		Resources? : Array<string>) 
		: Promise<SimplifiedSchema>
	{
		if(!Page) 
			Page = 0;
		let MaxResources = 500;
		
		let Request = xml('iq');
		Request.attrs.id = NewGuid();
		Request.attrs.type = "get";
		Request.attrs.to = this.DataSourceAddress;
		let Payload = xml( "read-schema", { "xmlns": LWTSD.Namespace });
		Payload.attrs.format = SchemaFormatToString(SchemaFormat.Simplified);
		Payload.attrs.startindex =  (Page * MaxResources).toString();
		Payload.attrs.maxresources = MaxResources.toString();

		if (AccessToken)
		{
			let AToken = xml( "accesstoken", { /*"xmlns": LWTSD.Namespace*/ }, AccessToken); 
			AToken.attrs.name = "urn:clayster:cdo:sessionid";
			Payload.append(AToken);
		}

		if (Resources)
		{
			for (let res of Resources)
			{
				let Resource = xml( "resource", { /*"xmlns": LWTSD.Namespace*/ }); 
				Resource.attrs.path = res;
				Payload.append(Resource);
			}
		}

		Request.append(Payload);

		let Response = await this.GetIQResponse(Request);

		let ReturnedData = Response.getChild("simplified-schema");
		if(!ReturnedData)
			throw new Error("Failed to get schema");

		return new SimplifiedSchema(ReturnedData);
	}

	private async InnerReadData(KnownSchema : SimplifiedSchema | null,
		ResourcePaths : Array<string>,
		Page? :number,
		ResubscriptionIDs? : Array<string> | null,
		AccessToken? : string|null)
	: Promise<DataPage>
	{
		if(!Page)
			Page = 0;
		if(!ResubscriptionIDs)
			ResubscriptionIDs = null;
		if(!AccessToken)
			AccessToken = null;
		if (KnownSchema == null)
			KnownSchema = await this.GetSchema(0, AccessToken, ResourcePaths);
		

		let Request = xml('iq');
		Request.attrs.id = NewGuid();
		Request.attrs.type = "get";
		Request.attrs.to = this.DataSourceAddress;

		let Payload = xml( "read-data", { "xmlns": LWTSD.Namespace });
		Payload.attrs.maxpoints = DataPage.PointsPerPage.toString();
		Payload.attrs.startindex = (Page * DataPage.PointsPerPage).toString();
		Payload.attrs.relativetimeout = "10";

		if (AccessToken != null)
		{
			let AToken = xml( "accesstoken", { /*"xmlns": LWTSD.Namespace*/ }, AccessToken); 
			AToken.attrs.name = "urn:clayster:cdo:sessionid";
			Payload.append(AToken);
		}

		if (ResubscriptionIDs)
		{
			for (let sid of ResubscriptionIDs)
			{
				let el = xml( "re-subscribe", { /*"xmlns": LWTSD.Namespace*/ }, AccessToken);
				el.attrs.subscriptionid = sid;
				Payload.append(el);
			}
		}

		for (let Path of ResourcePaths)
		{
			let el = xml( "read", { /*"xmlns": LWTSD.Namespace*/ }, AccessToken);
			el.attrs["resource-path"] = Path;
			el.attrs.maxpoints = "1";
			Payload.append(el);
		}

		Request.append(Payload);
		let Response = await this.GetIQResponse(Request);
		let ReturnedData = Response.getChild("data");

		if (!ReturnedData == null)
			throw new Error("Failed to get data: invalid data: " + Response.toString());


		// This is wrong for more than one or two iterations, Need to re-read the data. Recurse?
		let RVal = new DataPage(ReturnedData, KnownSchema, Page);
		return RVal;
	}


	public async ReadData(KnownSchema : SimplifiedSchema|null,
		ResourcePaths : Array<string>,
		AccessToken? :string|null,
		Page ? :number)
	: Promise<DataPage>
	{
		if(!Page)
			Page = 0;
		if(!AccessToken)
			AccessToken = null;

		let NrSchemaMissMatchRetries = 10;
		let SchemaMissMatched = true;

		for (let i = 0; i < NrSchemaMissMatchRetries && SchemaMissMatched; i++)
		{
			SchemaMissMatched = false;
			try
			{
				return await this.InnerReadData(KnownSchema, ResourcePaths, Page, new Array<string>(), AccessToken);
			}
			catch (Err)
			{
				if(Err as SchemaMissMatchException)
					SchemaMissMatched = true;

				throw new Error("Unhandled error in DataController : ReadData: " + Err.toString());
			}
		}

		throw new Error("Failed to read data" + (SchemaMissMatched ? ": schema miss matched" : ""));
	}

	public async  SubscribeToData(KnownSchema : SimplifiedSchema,
		ResourcePaths : Array<string>,
		NewData : (SubscriptionID: string, Data: DataPage)=>void,
		SubscriptionCancelled : (SubscriptionID: string)=>void,
		AccessToken?:string |null)
	: Promise<[string, DataPage]>
	{
		if(!AccessToken)
			AccessToken = null;
		
		let SubscriptionID = NewGuid();		
		let Payload = xml( "subscribe", { "xmlns": LWTSD.Namespace });
		Payload.attrs.subscriptionid = SubscriptionID;


		for (let res of ResourcePaths)
		{
			let Trigger = xml("trigger", { /*"xmlns": LWTSD.Namespace*/ });
			Trigger.attrs.onresource = res;
			Payload.append(Trigger);
		}

		let Request = xml('iq');
		Request.attrs.id = NewGuid();
		Request.attrs.type = "get";
		Request.attrs.to = this.DataSourceAddress;

		let NewSubscription = new Subscription();
		NewSubscription.ID = SubscriptionID;
		NewSubscription.NewData = NewData;

		Request.append(Payload);

		let Response = await this.GetIQResponse(Request);

		
		let Page = await this.ReadData(KnownSchema, ResourcePaths, AccessToken);
		NewSubscription.KnownSchema = Page.Schema;
		NewSubscription.Resources = ResourcePaths;
		if(AccessToken == null)
			NewSubscription.AccessToken = null;
		else
			NewSubscription.AccessToken = AccessToken as string;
		NewSubscription.SubscriptionCancelled = SubscriptionCancelled;

		this.Subscriptions.set(NewSubscription.ID, NewSubscription);

		throw new Error("DataController: SubscribeToData not implemented"); // Todo: must add invocation..
		//return [SubscriptionID, Page];
	}

	public async WriteData( Page : DataPage, AccessToken? : string  )
	: Promise<boolean>
	{			

		let Payload = xml( "write-data", { "xmlns": LWTSD.Namespace });
		if (AccessToken)
		{
			let AToken = xml( "accesstoken", { /*"xmlns": LWTSD.Namespace*/ }, AccessToken); 
			AToken.attrs.name = "urn:clayster:cdo:sessionid";
			Payload.append(AToken);
		}


		for (let res of Page.Data)
		{
			let write = xml( "write", { /*"xmlns": LWTSD.Namespace*/ }, res.GetWriteValue() ); 
			write.attrs["resource-path"] = res.Path;
			write.attrs.relativetimeout = "10";
			Payload.append(write);
		}

		let Request = xml('iq');
		Request.attrs.id = NewGuid();
		Request.attrs.type = "get";
		Request.attrs.to = this.DataSourceAddress;		
		Request.append(Payload);

		let Response = await this.GetIQResponse(Request);
		if(Response.attr.type = 'result')
			return true;

		return false;
	}


	private async HandleSubscriptionTriggered(Element : any)
	: Promise<boolean>
	{
		let SubscriptionID = Element.attrs.subscriptionid;
		if(!this.Subscriptions.has(SubscriptionID))
			return false;
		let Sub = this.Subscriptions.get(SubscriptionID);
		if(!Sub)
			return false;

		let Page = await this.InnerReadData(Sub.KnownSchema,
			Sub.Resources,
			0,
			[SubscriptionID],
			Sub.AccessToken);

		Sub.NewData(SubscriptionID, Page);
		Sub.LastReSubscription = new Date();
		return true;
	}

	public CancelSubscription(SubscriptionID : string) : void
	{
		if (!this.Subscriptions.has(SubscriptionID))
			return;

		this.Subscriptions.delete(SubscriptionID);

		let Request = xml("message");
		Request.attrs.to = this.DataSourceAddress;

		let Payload = xml( "cancel-subscription", { "xmlns": LWTSD.Namespace });
		Payload.attrs.subscriptionid = SubscriptionID;
		Request.append(Payload);
		this.Uplink.send(Request);
	}
	
	private HandleSubscriptionCancelled(CancelledElement : any) : void
	{
		let SubscriptionID = CancelledElement.attrs.subscriptionid;
		
		if (!this.Subscriptions.has(SubscriptionID))
		{
			return;
		}

		let Sub = this.Subscriptions.get(SubscriptionID);
		if(!Sub)
			return;
		
		this.Subscriptions.delete(SubscriptionID);

		Sub.SubscriptionCancelled(SubscriptionID);
	}


	public constructor( Uplink : any, DataSourceAddress : any )
	{
		this.Uplink = Uplink;
		this.DataSourceAddress = DataSourceAddress;
		this.Uplink.on('stanza', this.OnStanza.bind(this));
	}

}

File: NET/LWTSD/DataController.cs
using System;
using System.Xml;
using System.Xml.Linq;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using DXMPP;
using LWTSD.ResourceTypes;
using System.Threading;
using System.Threading.Tasks;

namespace LWTSD
{
	public class DataController : IDisposable
	{
		JID DataSourceAddress;
		Connection Uplink;

		ConcurrentDictionary<string, TaskCompletionSource<StanzaIQ>> IQRequests
			= new ConcurrentDictionary<string, TaskCompletionSource<StanzaIQ>>();

		class Subscription
		{
			public string ID;
			public Action<Tuple<string, DataPage>> NewData;
			public Action<string> SubscriptionCancelled;
			public List<ResourcePath> Resources;
			public SimplifiedSchema KnownSchema;
			public string AccessToken;
			public DateTime LastReSubscription = DateTime.UtcNow;
		}


		ConcurrentDictionary<string, Subscription> Subscriptions =
			new ConcurrentDictionary<string, Subscription>();


		public async Task<SimplifiedSchema> GetSchema(int Page = 0, string AccessToken = null, List<ResourcePath> Resources = null)
		{
			const int MaxResources = 500;

			TaskCompletionSource<StanzaIQ> Signal = new TaskCompletionSource<StanzaIQ>();

			StanzaIQ Request = new StanzaIQ(StanzaIQ.StanzaIQType.Get);
			Request.To = DataSourceAddress;
			XElement Payload = new XElement(LWTSD.Namespace + "read-schema");
			Payload.SetAttributeValue("format", SchemaFormat.Simplified.GetSerializedValue());
			Payload.SetAttributeValue("startindex", Page * MaxResources);
			Payload.SetAttributeValue("maxresources", MaxResources);

			if (AccessToken != null)
			{
				XElement AToken = new XElement(LWTSD.Namespace + "accesstoken");
				AToken.SetAttributeValue("name", "urn:clayster:cdo:sessionid");
				AToken.Value = AccessToken;
				Payload.Add(AToken);
			}

			if (Resources != null)
			{
				foreach (var res in Resources)
				{
					XElement Resource = new XElement(LWTSD.Namespace + "resource");
					Resource.SetAttributeValue("path", res);
					Payload.Add(Resource);
				}
			}

			Request.Payload.Add(Payload);
			IQRequests[Request.ID] = Signal;

			Uplink.SendStanza(Request);

			await Signal.Task;

			// Parse signal data

			if (Signal.Task.Result == null)
				throw new Exception("Failed to get schema");

			XElement ReturnedData = Signal.Task.Result.Payload.Element(LWTSD.Namespace + "simplified-schema");
			if (ReturnedData == null)
				throw new Exception("Failed to get schema");

			SimplifiedSchema RVal = new SimplifiedSchema(ReturnedData);

			return RVal;
		}


		public class DataPage
		{
			public const int PointsPerPage = 500;

			public class SchemaMissMatchException : System.Exception
			{
			}

			public List<Resource> Data = new List<Resource>();
			public SimplifiedSchema Schema;
			public int NrTotalPages;
			public int Page;

			public DataPage(XElement Element, SimplifiedSchema KnownSchema, int Page)
			{
				if (KnownSchema == null)
					throw new SchemaMissMatchException();

				string NewSchemaVersion = Element.Attribute("schemaversion").Value;
				if (NewSchemaVersion != KnownSchema.GetVersion())
					throw new SchemaMissMatchException();

				Schema = KnownSchema;
				this.NrTotalPages = (int)(((double)Convert.ToInt32(Element.Attribute("totalpoints").Value) + 0.5) / (double)PointsPerPage);


				foreach (var it in Element.Elements(LWTSD.Namespace + "resource"))
				{
					ResourcePath Path = it.Attribute("path").Value;
					switch (KnownSchema.Resources[Path].SimpleType)
					{
						case SimplifiedType.String:
							Data.Add(new ResourceString(it));
							break;
						case SimplifiedType.Integer:
							Data.Add(new ResourceInteger(it));
							break;
						case SimplifiedType.Base64Binary:
							Data.Add(new ResourceBase64Binary(it));
							break;
						case SimplifiedType.DateTime:
							Data.Add(new ResourceDateTime(it));
							break;
						case SimplifiedType.Boolean:
							Data.Add(new ResourceBoolean(it));
							break;
						case SimplifiedType.Decimal:
							Data.Add(new ResourceDecimal(it));
							break;
						case SimplifiedType.Duration:
							Data.Add(new ResourceDuration(it));
							break;
						case SimplifiedType.Float:
							Data.Add(new ResourceFloat(it));
							break;
						case SimplifiedType.Time:
							Data.Add(new ResourceTime(it));
							break;
                        case SimplifiedType.Double:
                            Data.Add(new ResourceDouble(it));
                            break;

                    }
                }
			}

			public DataPage()
			{
				Page = 0;
				NrTotalPages = 1;
				Schema = null;
			}
		}



		async Task<DataPage> InnerReadData(SimplifiedSchema KnownSchema,
										   List<ResourcePath> Resources,
										   int Page = 0,
										   List<string> ResubscriptionIDs = null,
										   string AccessToken = null)
		{
			if (KnownSchema == null)
				KnownSchema = await GetSchema(0, AccessToken, Resources);

			TaskCompletionSource<StanzaIQ> Signal = new TaskCompletionSource<StanzaIQ>();

			StanzaIQ Request = new StanzaIQ(StanzaIQ.StanzaIQType.Get);
			Request.To = DataSourceAddress;
			XElement Payload = new XElement(LWTSD.Namespace + "read-data");
			Payload.SetAttributeValue("maxpoints", DataPage.PointsPerPage.ToString());
			Payload.SetAttributeValue("startindex", (Page * DataPage.PointsPerPage).ToString());
			Payload.SetAttributeValue("relativetimeout", "10");

			if (AccessToken != null)
			{
				XElement AToken = new XElement(LWTSD.Namespace + "accesstoken");
				AToken.SetAttributeValue("name", "urn:clayster:cdo:sessionid");
				AToken.Value = AccessToken;
				Payload.Add(AToken);
			}

			if (ResubscriptionIDs != null)
			{
				foreach (string sid in ResubscriptionIDs)
				{
					XElement el = new XElement(LWTSD.Namespace + "re-subscribe");
					el.SetAttributeValue("subscriptionid", sid);
					Payload.Add(el);
				}
			}

			foreach (ResourcePath Path in Resources)
			{
				XElement el = new XElement(LWTSD.Namespace + "read");
				el.SetAttributeValue("resource-path", Path);
				el.SetAttributeValue("maxpoints", "1");
				Payload.Add(el);
			}

			Request.Payload.Add(Payload);
			IQRequests[Request.ID] = Signal;

			Uplink.SendStanza(Request);

			await Signal.Task;

			// Parse signal data

			if (Signal.Task.Result == null)
				throw new Exception("Failed to get data: no return value");

			XElement ReturnedData = Signal.Task.Result.Payload.Element(LWTSD.Namespace + "data");
			if (ReturnedData == null)
				throw new Exception("Failed to get data: invalid data: " + Signal.Task.Result.Payload.ToString());


			// This is wrong for more than one or two iterations, Need to re-read the data. Recurse?
			DataPage RVal = new DataPage(ReturnedData, KnownSchema, Page);
			return RVal;

		}


		public async Task<DataPage> ReadData(SimplifiedSchema KnownSchema,
											 List<ResourcePath> Resources,
											 string AccessToken = null,
											 int Page = 0)
		{
			const int NrSchemaMissMatchRetries = 10;
			bool SchemaMissMatched = true;
			for (int i = 0; i < NrSchemaMissMatchRetries && SchemaMissMatched; i++)
			{
				SchemaMissMatched = false;
				try
				{
					return await InnerReadData(KnownSchema, Resources, Page, new List<string>(), AccessToken);
				}
				catch (DataPage.SchemaMissMatchException)
				{
					SchemaMissMatched = true;
				}
			}
			throw new Exception("Failed to read data" + (SchemaMissMatched ? ": schema miss matched" : ""));
		}

		public async Task<Tuple<string,DataPage>> SubscribeToData(SimplifiedSchema KnownSchema,
		                                                          List<ResourcePath> Resources,
		                                                          Action<Tuple<string, DataPage>> NewDataAction,
		                                                          Action<string> SubscriptionCancelled,
		                                                          string AccessToken = null)
		{
			string SubscriptionID = Guid.NewGuid().ToString();

			XElement Payload = new XElement(LWTSD.Namespace + "subscribe");
			Payload.SetAttributeValue("subscriptionid", SubscriptionID);

			foreach (ResourcePath res in Resources)
			{
				XElement Trigger = new XElement(LWTSD.Namespace + "trigger");
				Trigger.SetAttributeValue("onresource", res);
				Payload.Add(Trigger);
			}
			StanzaIQ Request = new StanzaIQ(StanzaIQ.StanzaIQType.Set);
			Request.To = DataSourceAddress;

			Subscription NewSubscription = new Subscription();

			NewSubscription.ID = SubscriptionID;
			NewSubscription.NewData = NewDataAction;

			TaskCompletionSource<StanzaIQ> Signal = new TaskCompletionSource<StanzaIQ>();
			Request.Payload.Add(Payload);
			IQRequests[Request.ID] = Signal;

			Uplink.SendStanza(Request);

			await Signal.Task;

			if (Signal.Task.Result == null)
				throw new Exception("Failed to subscribe: no data");
			if (Signal.Task.Result.IQType != StanzaIQ.StanzaIQType.Result)
				throw new Exception("Subscription failed");


			DataPage Page = await ReadData(KnownSchema, Resources, AccessToken);
			NewSubscription.KnownSchema = Page.Schema;
			NewSubscription.Resources = Resources;
			NewSubscription.AccessToken = AccessToken;
			NewSubscription.SubscriptionCancelled = SubscriptionCancelled;

			Subscriptions[NewSubscription.ID] = NewSubscription;

			return new Tuple<string, DataPage>( SubscriptionID, Page);
		}

		public async Task<bool> WriteData(DataPage Page, string AccessToken = null)
		{			
			XElement Payload = new XElement( LWTSD.Namespace + "write-data" );

			if (AccessToken != null)
			{
				XElement AToken = new XElement(LWTSD.Namespace + "accesstoken");
				AToken.SetAttributeValue("name", "urn:clayster:cdo:sessionid");
				AToken.Value = AccessToken;
				Payload.Add(AToken);
			}

			foreach (var res in Page.Data)
			{
				XElement write = new XElement(LWTSD.Namespace + "write");
				write.SetAttributeValue("resource-path", res.Path);
				write.SetAttributeValue("relativetimeout", "10");
				write.Value = res.GetWriteValue();
				Payload.Add(write);
			}

			StanzaIQ Request = new StanzaIQ(StanzaIQ.StanzaIQType.Set);
			Request.To = DataSourceAddress;
			TaskCompletionSource<StanzaIQ> Signal = new TaskCompletionSource<StanzaIQ>();
			Request.Payload.Add(Payload);
			IQRequests[Request.ID] = Signal;

			Uplink.SendStanza(Request);

			await Signal.Task;

			if (Signal.Task.Result == null)
				throw new Exception("Failed to write data: result is null");
			if (Signal.Task.Result.IQType == StanzaIQ.StanzaIQType.Result)
				return true;

			return false;
		}

		async Task<bool> HandleSubscriptionTriggered(XElement Element)
		{
			string SubscriptionID = Element.Attribute("subscriptionid").Value;
			Subscription Sub = Subscriptions[SubscriptionID];
			if (Sub == null)
				return false;

			DataPage Page = await InnerReadData(Sub.KnownSchema,
												Sub.Resources,
												0,
												new List<String>(new string[] { SubscriptionID }),
			                                    Sub.AccessToken);

			var t = Task.Run(() => Sub.NewData.Invoke(new Tuple<string, DataPage>(SubscriptionID, Page)));
			Sub.LastReSubscription = DateTime.UtcNow;
			return true;
		}

		public void CancelSubscription(string SubscriptionID)
		{
			Subscription Sub;
			if (!Subscriptions.ContainsKey(SubscriptionID))
				return;

			Subscriptions.TryRemove(SubscriptionID, out Sub);

			StanzaMessage Request = new StanzaMessage();
			Request.To = DataSourceAddress;
			XElement Payload = new XElement(LWTSD.Namespace + "cancel-subscription");
			Payload.SetAttributeValue("subscriptionid", Sub.ID);
			Request.Payload.Add(Payload);
			Uplink.SendStanza(Request);
		}

		public DataController( Connection Uplink, JID DataProducerAddress )
		{
			this.Uplink = Uplink;
			this.DataSourceAddress = DataProducerAddress;
			this.Uplink.OnStanzaIQ += Uplink_OnStanzaIQ;
			this.Uplink.OnStanzaMessage += Uplink_OnStanzaMessage;
		}

		void Uplink_OnStanzaIQ(StanzaIQ Data)
		{
			if (Data.From.GetBareJID() != DataSourceAddress.GetBareJID())
				return;

			Console.WriteLine("Controller received iq stanza");
			
			if (!IQRequests.ContainsKey(Data.ID))
				return;
			TaskCompletionSource<StanzaIQ> Signal;
			if (!IQRequests.TryRemove(Data.ID, out Signal))
				return;

			Signal.SetResult(Data);
		}

		void HandleSubscriptionCancelled(XElement CancelledElement)
		{
			string SubscriptionID = CancelledElement.Attribute("subscriptionid").Value;
			Subscription Sub = null;
			Subscriptions.TryRemove(SubscriptionID, out Sub);
			if (Sub == null)
				return;

			Sub.SubscriptionCancelled.Invoke(SubscriptionID);
		}

		void Uplink_OnStanzaMessage(StanzaMessage Data)
		{
			if (Data.From.GetBareJID() != DataSourceAddress.GetBareJID())
				return;
			
			Console.WriteLine("Controller received message stanza");
			XElement LWTSDElement = Data.Payload.Elements().First();

			if (LWTSDElement.Name.Namespace != LWTSD.Namespace)
				return;

			switch (LWTSDElement.Name.LocalName)
			{
				case "subscription-cancelled":
					HandleSubscriptionCancelled(LWTSDElement);
					break;
				case "subscription-triggered":
					HandleSubscriptionTriggered(LWTSDElement);
					break;
			}

		}

		public void Dispose()
		{
        	Uplink.OnStanzaIQ -= Uplink_OnStanzaIQ;
			Uplink.OnStanzaMessage -= Uplink_OnStanzaMessage;
			Uplink = null;
		}
	}
}
